protofsm/state_machine.go

```diff
@@ -0,0 +1,428 @@
+}
+
+// StateSubscriber represents an active subscription to be notified of new
+// state transitions.
+type StateSubscriber[E any, F Environment] *fn.EventReceiver[State[E, F]]
+
+// RegisterStateEvents registers a new event listener that will be notified of
+// new state transitions.
+func (s *StateMachine[Event, Env]) RegisterStateEvents() StateSubscriber[Event, Env] {
+	subscriber := fn.NewEventReceiver[State[Event, Env]](10)
+
+	// TODO(roasbeef): instead give the state and the input event?
+
+	s.newStateEvents.RegisterSubscriber(subscriber)
+
+	return subscriber
+}
+
+// RemoveStateSub removes the target state subscriber from the set of active
+// subscribers.
+func (s *StateMachine[Event, Env]) RemoveStateSub(sub StateSubscriber[Event, Env]) {
+	s.newStateEvents.RemoveSubscriber(sub)
+}
+
+// executeDaemonEvent executes a daemon event, which is a special type of event
+// that can be emitted as part of the state transition function of the state
+// machine. An error is returned if the type of event is unknown.
+func (s *StateMachine[Event, Env]) executeDaemonEvent(event DaemonEvent) error {
+	switch daemonEvent := event.(type) {
+
+	// This is a send message event, so we'll send the event, and also mind
+	// any preconditions as well as post-send events.
+	case *SendMsgEvent[Event]:
+		sendAndCleanUp := func() error {
+			err := s.daemon.SendMessages(
+				daemonEvent.TargetPeer, daemonEvent.Msgs,
+			)
+			if err != nil {
+				return fmt.Errorf("unable to send msgs: %w", err)
+			}
+
+			// If a post-send event was specified, then we'll
+			// funnel that back into the main state machine now as
+			// well.
+			daemonEvent.PostSendEvent.WhenSome(func(event Event) {
+				s.wg.Add(1)
+				go func() {
+					defer s.wg.Done()
+
+					s.SendEvent(event)
+				}()
+			})
+
+			return nil
+		}
+
+		// If this doesn't have a SendWhen predicate, then we can just
+		// send it off right away.
+		if !daemonEvent.SendWhen.IsSome() {
+			if err := sendAndCleanUp(); err != nil {
+				return err
+			}
+		}
+
+		// Otherwise, this has a SendWhen predicate, so we'll need
+		// launch a goroutine to poll the SendWhen, then send only once
+		// the predicate is true.
+		s.wg.Add(1)
+		go func() {
+			defer s.wg.Done()
+
+			predicateTicker := time.NewTicker(pollInterval)
+			defer predicateTicker.Stop()
+
+			for {
+				select {
+				case <-predicateTicker.C:
+					canSend := fn.MapOptionZero(
+						daemonEvent.SendWhen,
+						func(pred SendPredicate) bool {
+							return pred()
+						},
+					)
+
+					if canSend {
+						sendAndCleanUp()
+						return
+					}
+
+				case <-s.quit:
+					return
+				}
+			}
+		}()
+
+		return nil
+
+	// If this is a disable channel event, then we'll disable the channel.
+	// This is usually done for things like co-op closes.
+	case *DisableChannelEvent:
+		err := s.daemon.DisableChannel(daemonEvent.ChanPoint)
+		if err != nil {
+			return fmt.Errorf("unable to disable channel: %w", err)
+		}
+
+		return nil
```

Comment:
> Are all of these state machines guaranteed to be attached to a particular channel id? This Event feels less general than the other two, but I don't have a concrete argument for why that's the case.

Yeah good point. In terms of usage, I guess this could become a new interface that we accept as part of the `Environment`. This one felt more syscall-y to me though, as the other interfaces don't affect/mutate the outside world, they just want to examine an attribute (no dangling updates), or do something pure like sign a signature.